viewer/src/logic/commands.ts

```diff
@@ -156,7 +157,13 @@ export function advanceResearchWarning(player: Player, field: ResearchField): Bu
   const warnings = events
     .filter((e) => e.operator == Operator.Once)
     .flatMap((e) => e.rewards)
-    .flatMap((r) => rewardWarnings(player, r));
+    .flatMap((r) => {
+      if (field == ResearchField.Science && level == 4 && r.type == Resource.Knowledge) {
+        // have to research first
+        r.count -= new Reward(UPGRADE_RESEARCH_COST).count;
+      }
+      return rewardWarnings(player, r);
+    });
```

Comment:
You can merge rewards.

```
const rewards =  Reward.merge(
    events.filter((e) => e.operator == Operator.Once).map(e => e.rewards), 
    Reward.negative(Reward.parse(UPGRADE_RESEARCH_COST));
const warnings = rewards.map(r => rewardWarnings(r, player));
```

This has the advantage of merging all the rewards / costs.

There is still a problem: sometimes the research is free (firakes, tech tile), some times it costs 4k. 

The engine should forward the cost of upgrade research in addition to the tracks in `available-commands`. That way we can directly use that cost instead of a hard-coded value.

Reply:
it was already in the available command..